rust/lance-index/src/scalar/bitmap.rs

```diff
@@ -0,0 +1,283 @@
+    PhysicalSortExpr,
+};
+use deepsize::DeepSizeOf;
+use futures::{
+    future::BoxFuture,
+    stream::{self},
+    FutureExt, Stream, StreamExt, TryFutureExt, TryStreamExt,
+};
+use lance_core::{Error, Result};
+use lance_datafusion::{
+    chunker::chunk_concat_stream,
+    exec::{execute_plan, LanceExecutionOptions, OneShotExec},
+};
+use roaring::RoaringBitmap;
+use roaring::treemap::RoaringTreemap;
+use serde::{Serialize, Serializer};
+use snafu::{location, Location};
+
+use crate::{Index, IndexType};
+
+use super::{
+    btree::{BTreeSubIndex, BtreeTrainingSource}, flat::FlatIndexMetadata, IndexReader, IndexStore, IndexWriter, ScalarIndex, ScalarQuery
+};
+use std::collections::HashSet;
+
+use super::btree::OrderableScalarValue;
+use bytes::Bytes;
+
+
+const BITMAP_LOOKUP_NAME: &str = "bitmap_page_lookup.lance";
+const BITMAP_PAGES_NAME: &str = "bitmap_page_data.lance";
+
+#[derive(Clone, Debug)]
+pub struct LanceRoaringTreemap(RoaringTreemap);
+
+impl DeepSizeOf for LanceRoaringTreemap {
+    fn deep_size_of_children(&self, context: &mut deepsize::Context) -> usize {
+        0
+    }
+}
+
+#[derive(Clone, Debug, DeepSizeOf)]
+pub struct BitmapIndex {
+    index_map: BTreeMap<OrderableScalarValue, LanceRoaringTreemap>,
+    page_map: BTreeMap<OrderableScalarValue, Vec<u64>>,
+    store: Arc<dyn IndexStore>,
+}
+
+impl BitmapIndex {
+    fn new(
+        index_map: BTreeMap<OrderableScalarValue, LanceRoaringTreemap>,
+        page_map: BTreeMap<OrderableScalarValue, Vec<u64>>,
+        store: Arc<dyn IndexStore>,
+    ) -> Self {
+        Self {
+            index_map,
+            page_map,
+            store,
+        }
+    }
+
+    // creates a new BitmapIndex from a serialized RecordBatch
+    fn try_from_serialized(data: RecordBatch, store: Arc<dyn IndexStore>) -> Result<Self> {
+
+        if data.num_rows() == 0 {
+            return Err(Error::Internal {
+                message: "attempt to load bitmap index from empty record batch".into(),
+                location: location!(),
+            });
+        }
+
+        let dict_keys = data.column(0);
+        let binary_bitmaps = data.column(1);
+        let pages_column = data.column(2);
+        let bitmap_binary_array = binary_bitmaps.as_any().downcast_ref::<BinaryArray>().unwrap();
+        let pages_array = pages_column.as_any().downcast_ref::<ListArray>().unwrap();
+
+        let mut index_map: BTreeMap<OrderableScalarValue, LanceRoaringTreemap> = BTreeMap::new();
+        let mut page_map: BTreeMap<OrderableScalarValue, Vec<u64>> = BTreeMap::new();
+
+        for idx in 0..data.num_rows() {
+            let key = OrderableScalarValue(ScalarValue::try_from_array(dict_keys, idx)?);
+            let bitmap_bytes = bitmap_binary_array.value(idx);
+            let bitmap = LanceRoaringTreemap(RoaringTreemap::deserialize_from(bitmap_bytes).unwrap());
+
+            let page_numbers = pages_array.value(idx).as_any().downcast_ref::<UInt64Array>().unwrap().values().to_vec();
```

Comment:
Is there a good reason to use a vec instead of just reusing the `UInt64Array`? `to_vec()` will require copying the buffer, which seems unnecessary.